File: app/src/main/java/com/culture/santos/module/State.java
package com.culture.santos.module;

import android.util.Log;

import com.culture.santos.culture.MapsActivity;

import java.util.Observable;

/**
 * Created by Ricar on 02/09/2016.
 */
public class State extends Observable {

    public enum States{
        InitState,
        AddEvent,
        RemoveState,
        ListState,
        EditState
    }

    public enum StatesTutorial{
        Disable,
        AddEvent
    }

    private States currentState;
    private StatesTutorial currentStateTutorial;

    private StateManager stateManager;
    private MapsActivity cxt;

    public States getState(){
        return currentState;
    }
    public StatesTutorial getStateTutorial(){
        return currentStateTutorial;
    }

    public State(MapsActivity initContext){
        cxt = initContext;
        stateManager = new StateManager(initContext, this);
        setInitState();
    }

    public void setAddEventState(){
        if(isAddEventStateTutorial()) cxt.getTutorialAdapter().showAddEventTutorial();
        synchronized (this) {currentState = States.AddEvent;}
        notifyChanged();
    }

    public void setInitState(){
        synchronized (this) {
            currentState = States.InitState;
            currentStateTutorial = StatesTutorial.Disable;
        }
        notifyChanged();
    }

    public void setRemoveEventState() {
        synchronized (this) {currentState = States.RemoveState;}
        notifyChanged();
    }

    public void setEditEventState() {
        synchronized (this) {currentState = States.EditState;}
        notifyChanged();
    }

    public void setListEventState() {
        synchronized (this) {currentState = States.ListState;}
        notifyChanged();
    }

    public void setDisableTutorial(){
        synchronized (this) {currentStateTutorial = StatesTutorial.Disable;}
        notifyChanged();
    }

    public void setAddEventTutorial(){
        synchronized (this) {
            Log.d("BEFORE", currentStateTutorial.toString());
            if(isDisableEventStateTutorial()){
                currentStateTutorial = StatesTutorial.AddEvent;
            }else if(isAddEventStateTutorial()){
                currentStateTutorial = StatesTutorial.Disable;
            }
            Log.d("AFTER", currentStateTutorial.toString());
        }
        notifyChanged();
    }

    public boolean isAddEventStateTutorial(){return (currentStateTutorial == StatesTutorial.AddEvent);}

    public boolean isDisableEventStateTutorial(){return (currentStateTutorial == StatesTutorial.Disable);}

    public boolean isAddEventState(){
        return (currentState == States.AddEvent);
    }

    public boolean isRemoveEventState(){
        return (currentState == States.RemoveState);
    }

    private void notifyChanged(){
        setChanged();
        notifyObservers();
    }
}

File: app/src/main/java/com/culture/santos/module/State.kt
package com.culture.santos.module

import android.util.Log
import com.culture.santos.culture.MapsActivity
import java.util.Observable

/**
 * Created by Ricar on 02/09/2016.
 */
class State(private val cxt: MapsActivity?) : Observable() {

    var state: States? = null
        private set
    var stateTutorial: StatesTutorial? = null
        private set

    private val stateManager: StateManager

    val isAddEventStateTutorial: Boolean
        get() = stateTutorial == StatesTutorial.AddEvent

    val isDisableEventStateTutorial: Boolean
        get() = stateTutorial == StatesTutorial.Disable

    val isAddEventState: Boolean
        get() = state == States.AddEvent

    val isRemoveEventState: Boolean
        get() = state == States.RemoveState

    enum class States {
        InitState,
        AddEvent,
        RemoveState,
        ListState,
        EditState
    }

    enum class StatesTutorial {
        Disable,
        AddEvent
    }

    init {
        stateManager = StateManager(cxt, this)
        setInitState()
    }

    fun setAddEventState() {
        if (isAddEventStateTutorial) cxt?.tutorialAdapter!!.showAddEventTutorial()
        synchronized(this) {
            state = States.AddEvent
        }
        notifyChanged()
    }

    fun setInitState() {
        synchronized(this) {
            state = States.InitState
            stateTutorial = StatesTutorial.Disable
        }
        notifyChanged()
    }

    fun setRemoveEventState() {
        synchronized(this) {
            state = States.RemoveState
        }
        notifyChanged()
    }

    fun setEditEventState() {
        synchronized(this) {
            state = States.EditState
        }
        notifyChanged()
    }

    fun setListEventState() {
        synchronized(this) {
            state = States.ListState
        }
        notifyChanged()
    }

    fun setDisableTutorial() {
        synchronized(this) {
            stateTutorial = StatesTutorial.Disable
        }
        notifyChanged()
    }

    fun setAddEventTutorial() {
        synchronized(this) {
            Log.d("BEFORE", stateTutorial!!.toString())
            if (isDisableEventStateTutorial) {
                stateTutorial = StatesTutorial.AddEvent
            } else if (isAddEventStateTutorial) {
                stateTutorial = StatesTutorial.Disable
            }
            Log.d("AFTER", stateTutorial!!.toString())
        }
        notifyChanged()
    }

    private fun notifyChanged() {
        setChanged()
        notifyObservers()
    }
}
